op-node/rollup/sync/start.go

```diff
@@ -40,164 +27,192 @@ import (
+// and the same holds for all its ancestors.
+func FindL2Heads(ctx context.Context, cfg *rollup.Config, l1 L1Chain, l2 L2Chain) (result *FindHeadsResult, err error) {
+	// Fetch current L2 forkchoice state
+	result, err = currentHeads(ctx, cfg, l2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch current L2 forkchoice state: %w", err)
+	}
 
-	// Loop 1. Walk the L2 chain backwards until we find an L2 block whose L1 origin is canonical.
+	// Remember original unsafe block to determine reorg depth
+	prevUnsafe := result.Unsafe
 
 	// Current L2 block.
-	n := start
-
-	// Number of blocks between n and start.
-	reorgDepth := 0
-
-	// Blockhash of L1 origin hash for the L2 block during the previous iteration, 0 for first
-	// iteration. When this changes as we walk the L2 chain backwards, it means we're seeing a different
-	// (earlier) epoch.
-	var prevL1OriginHash common.Hash
+	n := result.Unsafe
 
-	// The highest L2 ancestor of `start` (or `start` itself) whose ancestors are not (yet) known
-	// to have a non-canonical L1 origin. Empty if no such candidate is known yet. Guaranteed to be
-	// set after exiting from Loop 1.
-	var highestPlausibleCanonicalOrigin eth.L2BlockRef
+	var highestL2WithCanonicalL1Origin eth.L2BlockRef
+	var l1Block eth.L1BlockRef
+	var ahead bool
 
 	for {
-		// Check if l1Origin is canonical when we get to a new epoch.
-		if prevL1OriginHash != n.L1Origin.Hash {
-			prevL1OriginHash = n.L1Origin.Hash
-
-			if plausible, canonical, err := isAheadOrCanonical(ctx, l1, n.L1Origin); err != nil {
-				return eth.L2BlockRef{}, eth.L2BlockRef{}, err
-			} else if !plausible {
-				// L1 origin nor ahead of L1 head nor canonical, discard previous candidate and
-				// keep looking.
-				highestPlausibleCanonicalOrigin = eth.L2BlockRef{}
-			} else {
-				if highestPlausibleCanonicalOrigin == (eth.L2BlockRef{}) {
-					// No highest plausible candidate, make L2 block new candidate.
-					highestPlausibleCanonicalOrigin = n
-				}
-				if canonical {
-					break
-				}
+		// Fetch L1 information if we never had it, of if we do not have it for the current origin
+		if l1Block == (eth.L1BlockRef{}) || n.L1Origin.Hash != l1Block.Hash {
+			b, err := l1.L1BlockRefByNumber(ctx, n.L1Origin.Number)
+			// if L2 is ahead of L1 view, then consider it a "plausible" head
+			notFound := errors.Is(err, ethereum.NotFound)
+			if err != nil && !notFound {
+				return nil, fmt.Errorf("failed to retrieve block %d from L1 for comparison against %s: %w", n.L1Origin.Number, n.L1Origin.Hash, err)
 			}
+			l1Block = b
+			ahead = notFound
 		}
 
-		// Don't walk past genesis. If we were at the L2 genesis, but could not find its L1 origin,
-		// the L2 chain is building on the wrong L1 branch.
-		if n.Hash == genesis.L2.Hash || n.Number == genesis.L2.Number {
-			return eth.L2BlockRef{}, eth.L2BlockRef{}, WrongChainErr
+		if n.Number == cfg.Genesis.L2.Number {
```

Comment:
why cut the comment?

Reply:
There are new more descriptive comments below within the if statement, and more descriptive errors too.

Reply:
Two reasons to keep it: it fits in the style of having a comment over each if stmt block. The second is that the reason we don't return on the case of `n == cfg.Genesis` is that `finalized = cfg.Genesis` in the case that we don't having a finalized block & we do a return on hitting `n == finalized` later on.